src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -439,42 +439,107 @@ internal abstract class Storage
 
         internal class ListStorage : Storage
         {
-            internal readonly List<KeyValuePair<string, object>> _inner;
+            private KeyValuePair<string, object>[] _items;
+            private int _size;
 
             public ListStorage()
             {
-                _inner = new List<KeyValuePair<string, object>>();
+                _items = new KeyValuePair<string, object>[0];
```

Comment:
Don't do this, keep a `readonly static` empty array and use that. The empty array is immutable so it can be shared. This is one of the many optimizations `List<T>` does.